BEST_PRACTICES.md

```diff
@@ -166,8 +160,98 @@ test:
+would be inconvenient if the Tap emitted all columns for all tables.
+
+To address this, we recommend allowing a Tap to produce a document
+containing the "discovered" schemas for its data source, and allowing it
+to also accept "annotated schemas" that indicate which streams and
+properties to sync.
+
+### Schema Discovery
+
+A Tap that wants to support property selection should add an optional
+`--discover` flag. When the `--discover` flag is supplied, the Tap should
+connect to its data source, find the list of streams available, and print
+out a document listing each stream along with the discovered schema. The
+discovery output should go to STDOUT, and it should be the only thing
+written to STDOUT. If the `--discovered` flag is supplied, a tap should
+not emit any RECORD, SCHEMA, or STATE messages.
+
+The format of the discovery output is as follows. The top level is an
+object, with a single key called "streams", that points to an array. Each
+item of the array is an object with two keys:
+
+* name - the name of the stream
+* schema - the discovered schema for the stream
+    
+The discovered schema is in JSON schema format, with one extension.
+Properties may optionally contain a `selectable` attribute, which means
+that a user can decide whether to include those properties in the output.
+Certain fields may not be selectable and will always be included if their
+parent object is included. For example, for a database source the primary
+key of each table must be included if the table is selected. An HTTP API
+might always emit some fields, while allowing other fields to be selected.
+
+Here's an example of a discovered schema:
+
+```javascript
+{"streams": [
+  {"name": "users",
+   "schema": {
+     "type": "object",
+     "properties": {
+       "id": {"type": "integer"},
+       "first_name": {"type": "string", "selectable": true},
+       "last_name": {"type": "string", "selectable": true},
+     }
+   }
+  },
+  {"name": "orders",
+   "schema": {
+     "type": "object",
+     "properties": {
+       "id": {"type": "integer"},
+       "user_id": {"type": "integer", "selectable": true},
+       "amount": {"type": "number", "selectable": true},
+       "credit_card_number": {"type": "string", "selectable": true},
+     }
+   }
+  }
+ ]
+}
+```
 
-
+2. Add `--properties PROPERTIES` option.
+
+    A tap that supports property selection should accept an optional
+    `--properties PROPERTIES` option. `PROPERTIES` should point to a file.
+    The Tap should limit its output to the streams present in the
+    `PROPERTIES` file, and further limit the fields in each stream to the
+    ones marked as "selected" in the PROPERTIES file.
+
+### Discovered Schema and Annotated Schema Format
+
+
+```javascript
+{"streams": [
+  {"name": "orders",
+   "schema": {
+     "type": "object",
+     "properties": {
+       "user_id": {"type": "integer", "selected": true},
+       "amount": {"type": "number", "selected": true},
```

Comment:
The `selected` flag here is redundant: Only selected fields are in the properties file, so that alone denotes that these fields are selected.